### PR TITLE
[WAX]: Add WAX chain

### PR DESCRIPTION
**Pull request by miloserdow**
## Description

Add WAX chain with basic transfer functions (#2670).

## How to test

Unit tests, check transaction on explorer

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [X] Reuse existing EOS code to add WAX
- [X] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description.

### CODE REVIEW

**Comment by SUPPORTTEAMHERR** [comment]
> ## Description
> 
> Add WAX chain with basic transfer functions (#2670).
> 
> ## How to test
> 
> Unit test
> 
> ## Types of changes
> 
> 
> 
> 
> 
> * New feature (non-breaking change which adds functionality)
> 
> 
> 
> ## Checklist
> 
> 
> 
> - [X] Create pull request as draft initially, unless its complete.
> - [X] Add tests to cover changes as needed.
> - [ ] Update documentation as needed.
> - [ ] If there is a related Issue, mention it in the description.
> 
> If you're adding a new blockchain
> 
> - [ ] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain.
> 

We'd be happy to look into this for you. Kindly file a support ticket here
https://forms.gle/zVGLP7GATPsDHdkH9

File: include/TrustWalletCore/TWCoinType.h [hunk 1]
```diff
@@ -43,6 +43,7 @@ enum TWCoinType {
     TWCoinTypeDigiByte = 20,
     TWCoinTypeDogecoin = 3,
     TWCoinTypeEOS = 194,
+    TWCoinTypeWAX = 194001,
```
Comment (catenocrypt) [inline include/TrustWalletCore/TWCoinType.h:46]: Where is this coinID coming from? We usually take coinIDs only from https://github.com/satoshilabs/slips/blob/master/slip-0044.md, to make sure that we use standard ones, and don't use a value which is used by someone else. But currently WAX is not present there. We should ask them to create a PR for it.

**Review by Milerius — CHANGES_REQUESTED** [review]
CI is failing, need to be fixed

**Comment by hewigovens** [comment]
If you run iOS and Android tests locally, you will find why it's failed, not random guess

**Comment by miloserdow** [comment]
> If you run iOS and Android tests locally, you will find why it's failed, not random guess

iOS worked locally, there's something weird going on, investigating...

**Comment by Milerius** [comment]
> > If you run iOS and Android tests locally, you will find why it's failed, not random guess
> 
> iOS worked locally, there's something weird going on, investigating...

There is nothing to investigate you just need to add few unit tests, this is very easy to spot if you compare with existing pull request that was merged adding a new chain:

- https://github.com/trustwallet/wallet-core/pull/2595/files#diff-409f71ddd7380158c6b0039185f287b61c5e7519cb3f39ef3dd5c4a89d189dbc
- https://github.com/trustwallet/wallet-core/pull/2595/files#diff-86476675b291daafb740aad7001dcde0757fb37d41bbcde777c33e959ae84616

File: include/TrustWalletCore/TWCoinType.h (hunk 1 above)
Comment (catenocrypt) [reply include/TrustWalletCore/TWCoinType.h:46]: If this is just 'internal' ID, I suggest using 10000194 (we use such 8 digits for internally made up IDs).

**Review by catenocrypt — APPROVED** [review]

File: include/TrustWalletCore/TWCoinType.h (hunk 1 above)
Comment (miloserdow) [reply include/TrustWalletCore/TWCoinType.h:46]: I just opened a pull request for slip-0044: https://github.com/satoshilabs/slips/pull/1473/files

Comment (miloserdow) [reply include/TrustWalletCore/TWCoinType.h:46]: Ok, it was merged, so now this is a legit value :)

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/CoinAddressDerivationTests.kt [hunk 2]
```diff
@@ -74,7 +74,7 @@ class CoinAddressDerivationTests {
         VIACOIN -> assertEquals("via1qnmsgjd6cvfprnszdgmyg9kewtjfgqflz67wwhc", address)
         QTUM -> assertEquals("QhceuaTdeCZtcxmVc6yyEDEJ7Riu5gWFoF", address)
         NULS -> assertEquals("NULSd6HgU8MoRnNjBgvJpa9tqvGxYdv5ne4en", address)
-        EOS -> assertEquals("EOS6hs8sRvGSzuQtq223zwJipMzqTJpXUVjyvHPvPwBSZWWrJTJkg", address)
+        EOS, WAX -> assertEquals("EOS6hs8sRvGSzuQtq223zwJipMzqTJpXUVjyvHPvPwBSZWWrJTJkg", address)
```
Comment (hewigovens) [inline android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/CoinAddressDerivationTests.kt:77]: Does wax still use `EOS` prefix?

**Review by Milerius — DISMISSED** [review]
Same question as @hewigovens - other than that seems good!

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/CoinAddressDerivationTests.kt (hunk 2 above)
Comment (miloserdow) [reply android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/CoinAddressDerivationTests.kt:77]: yes

File: tests/chains/WAX/TWAnySignerTests.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    const auto refBlock = parse_hex("0cffaeda15039f3468398c5b4295d220fcc217f7cf96030c3729773097c6bd76");
+    const auto key = parse_hex("d30d185a296b9591d648cb92fe0aa8f8a42de30ed9d2a21da9e7f69c67e8e355");
+
+    const auto pubKey = PublicKey(PrivateKey(key).getPublicKey(TWPublicKeyTypeSECP256k1));
+    const auto address = Address(pubKey);
+    EXPECT_EQ(address.string(), "EOS7rC6zYUjuxWkiokZTrwwHqwFvZ15Qdrn5WNxMKVXtHiDDmBWog");
+
+    auto& asset = *input.mutable_asset();
+    asset.set_amount(100000000);
+    asset.set_decimals(8);
```
Comment (catenocrypt) [inline tests/chains/WAX/TWAnySignerTests.cpp:32]: `registry.json` has 4 for decimals, not in sync

File: tests/chains/WAX/TWCoinTypeTests.cpp [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+// file LICENSE at the root of the source code distribution tree.
+
+#include "TestUtilities.h"
+#include <TrustWalletCore/TWCoinTypeConfiguration.h>
+#include <gtest/gtest.h>
+
+
+TEST(TWWAXCoinType, TWCoinType) {
+    auto symbol = WRAPS(TWCoinTypeConfigurationGetSymbol(TWCoinTypeWAX));
+    auto txId = WRAPS(TWStringCreateWithUTF8Bytes("testtx"));
```
Comment (hewigovens) [inline tests/chains/WAX/TWCoinTypeTests.cpp:14]: let's use real tx id and account

**Review by hewigovens — APPROVED** [review]

**Review by Milerius — APPROVED** [review]

**Review by catenocrypt — APPROVED** [review]

**Comment by catenocrypt** [comment]
So coinId `14001` is now present in Slip44 repo, initiated by our side :)
https://github.com/satoshilabs/slips/blob/master/slip-0044.md

**Pull request merged**